cpp/src/io/parquet/experimental/hybrid_scan_helpers.cpp

```diff
@@ -228,39 +243,238 @@ std::vector<std::vector<cudf::size_type>> aggregate_reader_metadata::filter_row_
   host_span<std::vector<cudf::size_type> const> row_group_indices,
   host_span<data_type const> output_dtypes,
   host_span<int const> output_column_schemas,
-  std::optional<std::reference_wrapper<ast::expression const>> filter,
+  std::reference_wrapper<ast::expression const> filter,
   rmm::cuda_stream_view stream) const
 {
-  // Return all row groups if no filter expression
-  if (not filter.has_value()) { return all_row_group_indices(row_group_indices); }
-
   // Compute total number of input row groups
-  cudf::size_type total_row_groups = [&]() {
-    if (not row_group_indices.empty()) {
-      size_t const total_row_groups =
-        std::accumulate(row_group_indices.begin(),
-                        row_group_indices.end(),
-                        size_t{0},
-                        [](auto sum, auto const& pfm) { return sum + pfm.size(); });
-
-      // Check if we have less than 2B total row groups.
-      CUDF_EXPECTS(total_row_groups <= std::numeric_limits<cudf::size_type>::max(),
-                   "Total number of row groups exceed the cudf::size_type's limit");
-      return static_cast<cudf::size_type>(total_row_groups);
-    } else {
-      return num_row_groups;
-    }
-  }();
+  auto const total_row_groups = compute_total_row_groups(row_group_indices);
 
   // Filter stats table with StatsAST expression and collect filtered row group indices
-  auto const stats_filtered_row_group_indices = apply_stats_filters(row_group_indices,
-                                                                    total_row_groups,
-                                                                    output_dtypes,
-                                                                    output_column_schemas,
-                                                                    filter.value(),
-                                                                    stream);
+  auto const stats_filtered_row_group_indices = apply_stats_filters(
+    row_group_indices, total_row_groups, output_dtypes, output_column_schemas, filter, stream);
 
   return stats_filtered_row_group_indices.value_or(all_row_group_indices(row_group_indices));
 }
 
+std::vector<byte_range_info> aggregate_reader_metadata::get_bloom_filter_bytes(
+  cudf::host_span<std::vector<cudf::size_type> const> row_group_indices,
+  host_span<data_type const> output_dtypes,
+  host_span<int const> output_column_schemas,
+  std::reference_wrapper<ast::expression const> filter)
+{
+  // Collect equality literals for each input table column
+  auto const literals =
+    equality_literals_collector{filter.get(), static_cast<cudf::size_type>(output_dtypes.size())}
+      .get_literals();
+
+  // Collect schema indices of columns with equality predicate(s)
+  std::vector<cudf::size_type> bloom_filter_col_schemas;
+  thrust::copy_if(thrust::host,
```

Comment:
`std::copy_if`?

Reply:
We need the `thrust` version to use the stencil on `literals`.

Reply:
I almost posted the same question :smile:

Reply:
Why `std::copy_if` cannot use `literals`? We can pass it as reference to the lambda predicate, right?